src/els_dt_document_index.erl

```diff
@@ -0,0 +1,97 @@
+%%==============================================================================
+%% The 'document_index' table
+%%==============================================================================
+
+-module(els_dt_document_index).
+
+%%==============================================================================
+%% Behaviour els_db_table
+%%==============================================================================
+
+-behaviour(els_db_table).
+-export([ name/0
+        , opts/0
+        ]).
+
+%%==============================================================================
+%% API
+%%==============================================================================
+
+-export([new/3]).
+
+-export([ find_by_kind/1
+        , insert/1
+        , lookup/1
+        ]).
+
+%%==============================================================================
+%% Includes
+%%==============================================================================
+-include("erlang_ls.hrl").
+
+%%==============================================================================
+%% Item Definition
+%%==============================================================================
+
+-record(els_dt_document_index, { id   :: els_dt_document:id()   | '_'
+                               , uri  :: uri()                  | '_'
+                               , kind :: els_dt_document:kind() | '_'
+                               }).
+-type els_dt_document_index() :: #els_dt_document_index{}.
+
+-type item() :: #{ id   := els_dt_document:id()
+                 , uri  := uri()
+                 , kind := els_dt_document:kind()
+                 }.
+-export_type([ item/0 ]).
+
+%%==============================================================================
+%% Callbacks for the els_db_table Behaviour
+%%==============================================================================
+
+-spec name() -> atom().
+name() -> ?MODULE.
+
+-spec opts() -> proplists:proplist().
+opts() ->
+  [ {attributes        , record_info(fields, els_dt_document_index)}
+  , {disc_copies       , [node()]}
+  , {index             , [#els_dt_document_index.kind]}
+  , {type              , bag}
```

Comment:
It seems like an overkill to have a bag to only support the fact the same id can be used for a module or a header. At this point, I would rather include the extension in the `id` (e.g. `file.hrl`) and use a `set`. Also this makes me think. Headers could indeed have the same `id` (independently of us including or not the extension), but different uris. In fact, whilst having two modules with clashing name is not typical, this can definitely happen for header files, sharing the same name but belonging to different apps. How do we plan to handle them?

Reply:
For headers with the same `id` and different `uri` there will be two different entries in the table since it's a `bag`. The same would happen with entries of type `other`.

> How do we plan to handle them?
We need to be mindful when using this index table of what we are trying to get and the fact that it's a table.